src/mtev_http.c

```diff
@@ -1194,56 +1309,114 @@ mtev_http_session_req_consume(mtev_http_session_ctx *ctx,
         ctx->req.next_chunk_read = 0;
         ctx->req.next_chunk = needed;
       }
-      len = blen;
+      user_len = blen;
     }
-    expected = ctx->req.next_chunk - ctx->req.next_chunk_read;
-    mtevL(http_debug, " ... need to read %d/%d more of a chunk\n", (int)expected,
+    wire_len = ctx->req.next_chunk - ctx->req.next_chunk_read;
+    mtevL(http_debug, " ... need to read %d/%d more of a chunk\n", (int)wire_len,
           (int)ctx->req.next_chunk);
   }
   mtevL(http_debug, " ... mtev_http_session_req_consume(%d) %d of %d\n",
-        ctx->conn.e ? ctx->conn.e->fd : -1, (int)len, (int)expected);
-  len = MIN(len, expected);
-  while(bytes_read < len) {
-    int crlen = 0;
+        ctx->conn.e ? ctx->conn.e->fd : -1, (int)user_len, (int)wire_len);
+
+  int crlen = 0;
+  while(bytes_read < user_len) {
     in = ctx->req.first_input;
-    while(in && bytes_read < len) {
-      int partial_len = MIN(in->size, len - bytes_read);
-      if(buf) memcpy((char *)buf+bytes_read, in->buff+in->start, partial_len);
-      bytes_read += partial_len;
-      ctx->req.content_length_read += partial_len;
-      if(ctx->req.payload_chunked) ctx->req.next_chunk_read += partial_len;
-      mtevL(http_debug, " ... filling %d bytes (read through %d/%d)\n",
-            (int)bytes_read, (int)ctx->req.content_length_read,
-            (int)ctx->req.content_length);
-      in->start += partial_len;
-      in->size -= partial_len;
-      if(in->size == 0) {
-        tofree = in;
-        ctx->req.first_input = in = in->next;
-        tofree->next = NULL;
-        RELEASE_BCHAIN(tofree);
-        if(in == NULL) {
-          ctx->req.last_input = NULL;
-          if (bytes_read != 0) {
-            mtevL(http_debug, " ... mtev_http_session_req_consume = %d\n",
-                  (int)bytes_read);
-            return bytes_read;
+    
+    while(in && in->size && bytes_read < user_len) {
```

Comment:
I see you did this by having the compressed and uncompressed data share a bchain... I think this doesn't behave as well as we'd like. I'm thinking of this scenario:
 * try to read compressed data - it allocates a bchain w/ compressed input, and appends an uncompressed bchain on successful decompression.
 * We did not fully drain the compressed buffer as we read uncompressed data.
 * We did not fully drain the uncompressed buffer (we didn't ask for that much data).
Then the next time we try to read, we still have uncompressed data available, but we won't notice it until we finish decompressing the data from the input buffer?